Adafruit_FreeTouch.cpp

```diff
@@ -94,10 +99,104 @@ bool Adafruit_FreeTouch::begin(void) {
+// calls with their low-level equivilents. Funny enough, the Generic Clock
+// Generator 3 is not actually used by arduinio core although it is set up in
+// various places.
+
+
+  const unsigned int PTC_CLOCK_GCGID = 3;
+
+  // Added for PTC - rough hack - JAG
+
+  SYSCTRL->OSC8M.bit.PRESC = SYSCTRL_OSC8M_PRESC_2_Val ;  // recent versions of CMSIS from Atmel changed the prescaler defines
+  SYSCTRL->OSC8M.bit.ONDEMAND = 1 ;
+  //  SYSCTRL->OSC8M.bit.RUNINSTANDBY = 1 ;
+
+  /* Put OSC8M as source for Generic Clock Generator 3 */
+  GCLK->GENDIV.reg = GCLK_GENDIV_ID( PTC_CLOCK_GCGID) ; // Generic Clock Generator 3
+  while ( (SYSCTRL->PCLKSR.reg & SYSCTRL_PCLKSR_DFLLRDY) == 0 );
+
+  // This magic actually does the assignment
+  GCLK->GENCTRL.reg =
+    ( GCLK_GENCTRL_ID( PTC_CLOCK_GCGID ) |
+      GCLK_GENCTRL_SRC_OSC8M
+      | GCLK_GENCTRL_GENEN
+    );
+  while ( (SYSCTRL->PCLKSR.reg & SYSCTRL_PCLKSR_DFLLRDY) == 0 );
+
+  // Original ASF code
+
+  // const int GCLK_GENERATOR_3 = 3;
+  // struct system_gclk_chan_config gclk_chan_conf;
+
+  // system_gclk_chan_get_config_defaults(&gclk_chan_conf);
+  // gclk_chan_conf->source_generator = GCLK_GENERATOR_0;
+  // equivilent to:
+  //  gclk_chan_conf.source_generator = GCLK_GENERATOR_3;
+
+  //  system_gclk_chan_set_config(PTC_GCLK_ID, &gclk_chan_conf);
+  // equivalent to
+  {
+    uint32_t new_clkctrl_config = (PTC_GCLK_ID << GCLK_CLKCTRL_ID_Pos);
+    new_clkctrl_config |= PTC_CLOCK_GCGID << GCLK_CLKCTRL_GEN_Pos;
+
+    //  system_gclk_chan_disable(channel);
+    // this is the "disable" function
+    {
+      /* Select the requested generator channel */
+      *((uint8_t*)&GCLK->CLKCTRL.reg) = PTC_CLOCK_GCGID;
+
+      /* Switch to known-working source so that the channel can be disabled */
+      uint32_t prev_gen_id = GCLK->CLKCTRL.bit.GEN;
+      GCLK->CLKCTRL.bit.GEN = 0;
+
+      /* Disable the generic clock */
+      GCLK->CLKCTRL.reg &= ~GCLK_CLKCTRL_CLKEN;
+      while (GCLK->CLKCTRL.reg & GCLK_CLKCTRL_CLKEN) {
+        /* Wait for clock to become disabled */
+      }
+
+      /* Restore previous configured clock generator */
+      GCLK->CLKCTRL.bit.GEN = prev_gen_id;
+    }
+
+    /* Write the new configuration */
+    GCLK->CLKCTRL.reg = new_clkctrl_config;
+  }
+
+  //  system_gclk_chan_enable(PTC_GCLK_ID);
+  // equivlent to:
+  *((uint8_t*)&GCLK->CLKCTRL.reg) = PTC_GCLK_ID;
+  GCLK->CLKCTRL.reg |= GCLK_CLKCTRL_CLKEN;
+
+
+  // system_apb_clock_set_mask(SYSTEM_CLOCK_APB_APBC, PM_APBCMASK_PTC);
+  // equivilant is:
+  PM->APBCMASK.reg |= PM_APBCMASK_PTC;
+
+#endif
+
+
+
+    // Ensure the pin is tri-stated with no-pullups (necessary for PTC to
+    // work).
+    //
+    // Note to developers: pinMode() appears to be the "safest" way to do
+    // peripheral selection because it ensures a single code-path inside
+    // ArduinoCore handles pinmux. Arduino "Input" mode is equivilent to
+    // peripheral "B", and this has been empirically confirmed.
+
+    pinMode( pin, INPUT );
```

Comment:
Thanks for all of the comments. I'll rework the PR and ping you again. This one point I wanted to f/u on. Was this library intended to somehow work inside of CircuitPython? Are there more details you can show me? The last read-through I made of the CP source code I recall seeing the the code to do it used the Atmel QTouch libraries.